urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/price/Price.java

```diff
@@ -0,0 +1,290 @@
+     */
+    public Price subtract(Price p) throws InvalidPriceOperation {
+        if (isMarket() == true || p.isMarket() == true) {
+            throw new InvalidPriceOperation("InvalidPriceOperation: Cannot subtract a LIMIT price from a MARKET Price");
+        } else {
+            long currentValue = getValue();
+            long subValue = p.VALUE;
+            long newValue = (currentValue - subValue);
+            Price newPrice = PriceFactory.makeLimitPrice(newValue);
+            return newPrice;
+        }
+    }
+
+    /**
+     * Public Multiply method which multiplies the current price by a number to
+     * get a new price.
+     *
+     * @param p Price object
+     * @return Price object
+     * @throws InvalidPriceOperation if the current price is a market Price or
+     * multiply value less than or equal to 0
+     * @see #getValue()
+     */
+    public Price multiply(int p) throws InvalidPriceOperation {
+        if (isMarket() == true) {
+            throw new InvalidPriceOperation("InvalidPriceOperation: Cannot multiply a MARKET price");
+        } else {
+            long currentValue = getValue();
+            long newValue = (currentValue * p);
+            Price newPrice = PriceFactory.makeLimitPrice(newValue);
+            return newPrice;
+        }
+    }
+
+    /**
+     * Public method which compares the current price to another price.
+     *
+     * @param p Price Object
+     * @return integer 1 if greater than, -1 if les than, 0 if equal
+     * @see #getValue()
+     */
+    @Override
+    public int compareTo(Price p) {
+        long comparable = p.VALUE;
+
+        if (getValue() == comparable) {
+            return 0;
+        } else if (getValue() < comparable) {
+            return -1;
+        } else if (getValue() > comparable) {
+            return 1;
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * Public method which compares two prices for greater than or equal to.
+     *
+     * @param p Price Object
+     * @return boolean if greater than or equal to comparable price
+     * @see #compareTo(price.Price)
+     * @see #isMarket()
+     */
+    public boolean greaterOrEqual(Price p) {
+        if (isMarket() || p.isMarket()) {
+            return false;
+        } else if (compareTo(p) == 0 || compareTo(p) == 1) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Public Greater Than method which compares if the current price is greater
+     * than another price.
+     *
+     * @param p Price object
+     * @return boolean true if greater than, false otherwise or if market price
+     * @see #isMarket()
+     * @see #compareTo(price.Price)
+     */
+    public boolean greaterThan(Price p) {
+        if (isMarket() || p.isMarket()) {
+            return false;
+        } else if (compareTo(p) == 1) {
+            return true;
+        } else {
+            return false;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 174:180, 213:219, 232:238.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.